README.md
### Please note this project is under heavy development. Expect sudden and breaking changes

![High Dimensional Resarch loves accessibility!](static/card.jpg)

# Introduction

High Dimensional Research lets agents use the internet. The internet was built for people, so HDR is creating tools to let agents browse, use, and learn from the internet. This is the just a client to make things easier for people building things.

Why this rules:

1. **Cheaper**: The internet has over 120 zettabytes of information on it. We only embed when pages change to save everyone a boat load of cash.
2. **Faster**: Can't beat a cache.
3. **Better**: Our embeddings are top notch since it is literally all we think about. We also leverage accessibility tooling and other neat tricks to isolate the information your agents are looking for.
4. **Greener**: GPUs use a lot of power. We save your wallet and the planet by creating an information commons.

If you have any questions, please email [Tynan Daly](mailto:tynan.daly@hdr.is)

## Usage

Please grab an api key from [hdr.is](https://hdr.is). Then either add it to your `.env` file or export it to your favorite shell via `export HDR_API_KEY=your-api-key-here`.

Clone the repo

```sh
git clone git@github.com:hdresearch/py-dimensional.git
```

Navigate to the directory

```sh
cd py-dimensional
```

Install `py-dimensional`

```
pip3 install .
```

Run examples

```sh
python examples/rag.py
python examples/summarize.py
```

## FAQ

> Is this yet another vector database?

No! HDR uses a new kind of vector index construction algorithm based on algebraic toplogy (we know, we can't believe an application actually exists either) to efficiently run vector search on the public internet. However, this means that we do not run search over your private documents. Go talk to our friends over at Pinecone, Milvus, or Chroma for that and let them know we sent you!

> Wait this is a browser?

Sort of. Static content is fetched from the vector cache and interactivity is done via headless browsers.

> What can I do with this?

Most people use the search function to help their agents navigate the internet and then switch over to the browser when the agents find what they're looking for.

Give them poetry, financial documents, real time news, etc. Run rag. Summarize text. Search for images (soon). Finally stop worrying about that damn training cutoff date.

> How do I singup? I dont see anything on your website.

[Sign up](https://re8zt94ow1u.typeform.com/to/qo0GQ398)

> Are my searches private?

No. We use search results to reweight the search indexes and (soon) customize them to your agents. Besides, someone has to make sure the robots behave.

> Do you have a discord?

No!

## Contribution

If you would like to contribute to this project:

1. Fork the project
2. Create your feature branch (`git checkout -b feature/AmazingFeature`)
3. Commit your changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the branch (`git push origin feature/AmazingFeature`)
5. Open a pull request

## License

This project is licensed under the Apache License 2.0 - see the [LICENSE](LICENSE) file for details.

## Testing

We use poetry for development :)

```bash
poetry install --with dev
pytest
```


CHANGELOG.md
# CHANGELOG

## 0.0.0

Welcome to HDR!


tests/test_hdr.py
import pytest
from dotenv import load_dotenv
import os

from py_dimensional.hdr import HDR

load_dotenv()
HDR_API_KEY = os.environ.get("HDR_API_KEY")
hdr = HDR(HDR_API_KEY)


class TestPage:
    url = "https://www.sec.gov/Archives/edgar/data/796343/000114036122033413/ny20005310x2_ex99-1.htm"

    def test_sec_page(self):
        page = hdr.page(self.url)
        assert page.url == self.url
        assert page.page_content[0:20] == "Exhibit 99.1\n\n![](ht"


class TestPageSearch:
    query = "How large is the figma aqusistion?"
    url = "https://www.sec.gov/Archives/edgar/data/796343/000114036122033413/ny20005310x2_ex99-1.htm"

    def test_page_search(self):
        results = hdr.page_search(self.url, self.query)
        assert results[0].content[:29] == "Investors and security holder"
        assert results[0].similarity > 0.8


class TestVectorSearch:
    query = "How large is the figma aqusistion?"

    def test_vector_search_engine(self):
        results = hdr.vector_search_engine(self.query)
        assert results[0].similarity > 0.8


py_dimensional/hdr.py
from urllib.parse import urljoin, urlencode
from .hdr_types import PageResponse, PageSearchResult
from typing import List, Dict
import httpx


class HDR:
    api_key: str
    base_url: str = "https://api.hdr.is/"
    api_version: str = "v1"
    api_endpoint: str
    headers: Dict[str, str]

    _client: httpx.Client
    _aclient: httpx.AsyncClient

    def __init__(self, api_key: str, **client_kwargs):
        super().__init__()
        self.api_key = api_key

        self.headers = self._construct_headers(self.api_key, uses_x_key=True)
        self.api_endpoint = urljoin(self.base_url, self.api_version)

        self._client = httpx.Client(**client_kwargs)
        self._aclient = httpx.AsyncClient(**client_kwargs)

    def _construct_headers(self, api_key: str, uses_x_key: bool = False):
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Accept": "application/json",
        }
        return headers

    def _construct_url(self, path: str, **kwargs):
        return f"{self.api_endpoint}/{path}"

    def page(self, url: str) -> PageResponse:
        response = self._client.get(
            self._construct_url("pageproxy"), params={"url": url}, headers=self.headers
        )
        if response.status_code == 200:
            return PageResponse(**response.json())
        else:
            return f"Request failed with status code {response.status_code}"

    def page_search(
        self,
        url: str,
        query: str,
    ) -> PageSearchResult:
        endpoint = self._construct_url("pagequery")
        response = self._client.get(
            endpoint, params={"url": url, "query": query}, headers=self.headers
        )
        if response.status_code == 200:
            return [PageSearchResult(**obj) for obj in response.json()]
        else:
            raise httpx._exceptions.HTTPError(
                f"Request failed with status code {response.status_code}"
            )

    def vector_search_engine(
        self, query: str, limit: int = 3
    ) -> List[PageSearchResult]:
        endpoint = self._construct_url("vectorsearch")
        response = self._client.get(
            endpoint,
            params={"query": query},
            headers=self.headers,
            timeout=1000,
        )

        if response.status_code == 200:
            return [PageSearchResult(**payload) for payload in response.json()]
        else:
            print(response.text)
            raise httpx._exceptions.HTTPError(
                f"Request failed with status code {response.status_code}"
            )


examples/summarization.py
import os
from dotenv import load_dotenv
from py_dimensional.hdr import HDR
import openai

load_dotenv()

HDR_API_KEY = os.environ.get("HDR_API_KEY")
hdr = HDR(HDR_API_KEY)
GPT_MODEL = "gpt-3.5-turbo-16k"


def page_summary(url: str):
    page_content = hdr.page(url).page_content
    completion = openai.ChatCompletion.create(
        model=GPT_MODEL,
        messages=[
            {
                "role": "user",
                "content": "Please summarize the following web page: " + page_content,
            },
        ],
        temperature=0.0,
    )
    text = completion.choices[0].message.content
    return text


url = "https://www.sec.gov/Archives/edgar/data/796343/000114036122033413/ny20005310x2_ex99-1.htm"
print(f"Source webpage: {url}")
print(f"Page summary: {page_summary(url)}")


examples/rag.py
import os
from dotenv import load_dotenv
from py_dimensional.hdr import HDR
import openai

load_dotenv()

HDR_API_KEY = os.environ.get("HDR_API_KEY")
GPT_MODEL = "gpt-3.5-turbo"

hdr = HDR(HDR_API_KEY)


def qa_over_webpage(url: str, query: str, limit: int = 3):
    results = hdr.page_search(url, query)

    context_str = ""
    for result in results[:limit]:
        context_str += result.content + "\n"

    completion = openai.ChatCompletion.create(
        model=GPT_MODEL,
        messages=[
            {
                "role": "system",
                "content": "Use the following context to answer the question: "
                + context_str,
            },
            {"role": "user", "content": query},
        ],
        temperature=0.0,
    )
    text = completion.choices[0].message.content
    return text


query = "How large is the figma aqusistion?"
url = "https://www.sec.gov/Archives/edgar/data/796343/000114036122033413/ny20005310x2_ex99-1.htm"


print(f"Source webpage: {url}")
print(
    f"""
    Question: {query}
    Answer: {qa_over_webpage(url, query)}
      """
)
query = "When is the conference call?"
print(
    f"""
    Question: {query}
    Answer: {qa_over_webpage(url, query)}
      """
)

query = "Who is buying figma?"
print(
    f"""
    Question: {query}
    Answer: {qa_over_webpage(url, query)}
    """
)


def rag_over_the_entire_internet(query: str, limit: int = 10):
    results = hdr.vector_search_engine(query)
    context_str = ""
    for result in results[:limit]:
        context_str += result.content + "\n"

    completion = openai.ChatCompletion.create(
        model=GPT_MODEL,
        messages=[
            {
                "role": "system",
                "content": "Use the following context to answer the question: "
                + context_str,
            },
            {"role": "user", "content": query},
        ],
        temperature=0.0,
    )
    text = completion.choices[0].message.content
    return text


query = "Who is buying figma?"
print(
    f"""
    Question: {query}
    Answer: {rag_over_the_entire_internet(query)}
"""
)
